digdag-core/src/main/java/io/digdag/core/database/DatabaseSecretControlStore.java

```diff
@@ -50,15 +51,81 @@ public void setProjectSecret(int projectId, String scope, String key, String val
     public void deleteProjectSecret(int projectId, String scope, String key)
     {
         transaction((handle, dao) -> {
-            dao.deleteProjectSecret(siteId, projectId, scope, key);
+            new LockedControl(handle, dao).deleteProjectSecret(projectId, scope, key);
             return null;
         });
     }
 
     @Override
     public List<String> listProjectSecrets(int projectId, String scope)
     {
-        return transaction((handle, dao) -> dao.listProjectSecrets(siteId, projectId, scope));
+        return transaction((handle, dao) -> {
+            return new LockedControl(handle, dao).listProjectSecrets(projectId, scope);
+        });
+    }
+
+    private class LockedControl
+            implements SecretControlStore
+    {
+        private final Handle handle;
+        private final Dao dao;
+
+        public LockedControl(Handle handle, Dao dao)
+        {
+            this.handle = handle;
+            this.dao = dao;
+        }
+
+        @Override
+        public void setProjectSecret(int projectId, String scope, String key, String value)
+        {
+            String encrypted = crypto.encryptSecret(value);
+            String engine = crypto.getName();
+
+            dao.upsertProjectSecret(siteId, projectId, scope, key, engine, encrypted);
+        }
+
+        @Override
+        public void deleteProjectSecret(int projectId, String scope, String key)
+        {
+            dao.deleteProjectSecret(siteId, projectId, scope, key);
+        }
+
+        @Override
+        public List<String> listProjectSecrets(int projectId, String scope)
+        {
+            return dao.listProjectSecrets(siteId, projectId, scope);
+        }
+
+        @Override
+        public <T> T lockProjectSecret(int projectId, String scope, String key, SecretLockAction<T> action)
+        {
+            EncryptedSecret secret = dao.lockProjectSecret(siteId, projectId, scope, key);
+
+            // TODO this logic is copied from DatabaseSecretStore.getSecret.
+            Optional<String> value;
+            if (secret == null) {
+                value = Optional.absent();
+            }
+            else {
+                // TODO: look up crypto engine using name
+                if (!crypto.getName().equals(secret.engine)) {
+                    throw new AssertionError("Crypto engine mismatch");
```

Comment:
Adding the values of `crypt.getName()` and `secret.engine` into the message would be better.